components/notion/sources/updated-page/updated-page.mjs

```diff
@@ -0,0 +1,51 @@
+import notion from "../../notion.app.mjs";
+import utils from "../common/utils.mjs";
+
+export default {
+  key: "notion-udpated-page",
+  name: "Updated Page", /* eslint-disable-line pipedream/source-name */
+  description: "Emit new event when a page is updated",
+  version: "0.0.1",
+  type: "source",
+  dedupe: "last",
+  props: {
+    db: "$.service.db",
+    timer: {
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 15,
+      },
+    },
+    notion,
+    databaseId: {
+      propDefinition: [
+        notion,
+        "databaseId",
+      ],
+    },
+  },
+  async run() {
+    const params = {
+      start_cursor: utils.getLastCursor(this.db),
+    };
+
+    do {
+      const response = await this.notion.queryDatabase(this.databaseId, params);
```

Comment:
Attempted to `filter` with `last_edited_time` and the results were the same as the comment above

Reply:
Currently, I think this source:
- fetches pages of items in the [API endpoint's](https://developers.notion.com/reference/post-database-query) default order
- [stores the last (_non-null_) `next_cursor`](https://github.com/PipedreamHQ/pipedream/blob/bc1a3fb348048b8a30634e6f789cebea9fe55992/components/notion/sources/updated-page/updated-page.mjs#L46-L48) from the retrieved pages to use as the starting cursor in the next run

The value of the cursor seems to be the `id` of the first item in the next page. If items returned by the request are ordered by `create_time`, then only events for items created after the cursor item will be emitted (since [`dedupe`](https://pipedream.com/docs/components/api/#dedupe-strategies) is [`last`](https://github.com/PipedreamHQ/pipedream/blob/bc1a3fb348048b8a30634e6f789cebea9fe55992/components/notion/sources/updated-page/updated-page.mjs#L10).
For example, if the `cursor` is the `id` of "Item 5", and the following updates are made:
1. Create "Item 6"
2. Edit "Item 2"
3. Edit "Item 6"
4. Create "Item 7"

I think `notion.databases.query({ database_id, start_cursor: cursor })` returns an array containing "Item 6" and "Item 7", missing an event for "Item 2".

Alternatively, if items are ordered by `last_edited_time`, and the cursor item is edited after other items are edited, the next request using that cursor will miss those items.
For example, if the `cursor` is the ID of "Item 5", and the following edits are made:
1. Edit "Item 2"
6. Edit "Item 9"
7. Edit "Item 5"
8. Edit "Item 3"

`notion.databases.query({ database_id, start_cursor: cursor })` returns an array containing "Item 5" and "Item 3", missing events for "Item 2" and "Item 9".

Also, if I delete the cursor item, I get an error on the next run of the source.
<img width="750" alt="error-start-cursor" src="https://user-images.githubusercontent.com/19861096/157552094-764dab77-1ef5-4454-92d3-627c1c017c13.png">

To fix this issue, I think we can either
- Fetch all items (pages) on each run of the source and filter out items with `last_edited_time` greater than the max `last_edited_time` from previous runs of the source
- Fetch items in order of `last_edited_time`, descending, and stop fetching when we see an item whose `last_edited_time` is >= the max `last_edited_time` from previous runs of the source. Reverse the list of items to emit in ascending time order. (e.g., [textlocal-new-sent-api-message](https://github.com/PipedreamHQ/pipedream/blob/master/components/textlocal/sources/new-sent-api-message/new-sent-api-message.js))

I'd suggest the second approach, to [reduce the number of API requests](https://pipedream.com/docs/components/guidelines/#rate-limit-optimization).

For example:
- `updated-page.mjs`
```js
import common from "../common/common.mjs";

const { notion } = common.props;

export default {
  ...common,
  key: "notion-udpated-page",
  name: "Updated Page", /* eslint-disable-line pipedream/source-name */
  description: "Emit new event when a page is updated",
  version: "0.0.1",
  type: "source",
  dedupe: "unique",
  props: {
    ...common.props,
    databaseId: {
      propDefinition: [
        notion,
        "databaseId",
      ],
    },
  },
  methods: {
    ...common.methods,
    _setLastEditedTime(lastEditedTime) {
      this.db.set("lastEditedTime", lastEditedTime);
    },
    _getLastEditedTime() {
      return this.db.get("lastEditedTime");
    },
    generateMeta(page) {
      const {
        last_edited_time: lastEditedTime,
        id,
      } = page;
      const title = this.notion.extractPageTitle(page);

      const ts = Date.parse(lastEditedTime);
      const compositeId = `${id}-${ts}`;
      return {
        // Create composite ID so update events for the same page have unique keys
        // See https://pipedream.com/docs/components/api/#emit
        id: compositeId,
        summary: `Page updated: ${title} - ${id}`,
        ts,
      };
    },
    isResultNew(result, startTimestamp) {
      const startDate = Date.parse(startTimestamp);
      const editedDate = Date.parse(result.last_edited_time);
      return editedDate > startDate;
    },
  },
  async run() {
    // If there is no lastEditedTime yet, include results edited after 7 days
    // ago so all pages aren't emitted on the first run
    const lastEditedTime = this._getLastEditedTime() ?? this.notion.daysAgo(7);

    const params = {
      sorts: [
        {
          timestamp: "last_edited_time",
          direction: "descending",
        },
      ],
    };

    // Get pages in updated order descending until the first page edited after
    // lastEditedTime, then reverse list of pages and emit

    const pagesStream = this.notion.getPages(this.databaseId, params);

    let pages = [];
    for await (const page of pagesStream) {
      if (!this.isResultNew(page, lastEditedTime)) {
        break;
      }
      pages.push(page);
    }

    pages.reverse().forEach((page) => {
      const meta = this.generateMeta(page);
      this.$emit(page, meta);
    });

    const lastPageEditedTime = pages[pages.length - 1]?.last_edited_time;
    if (lastPageEditedTime) {
      this._setLastEditedTime(lastPageEditedTime);
    }
  },
};
```

- In `notion.app.mjs`
```js
  methods: {
    // ...
    /**
     * This generator function scans the pages in a database yields each page
     * separately.
     *
     * @param {string} databaseId - The database containing the pages to scan
     * @param {object} [opts] - Options to customize the operation
     * @yield {object} The next page
     */
    async *getPages(databaseId, opts = {}) {
      let cursor;

      do {
        const params = {
          ...opts,
          start_cursor: cursor,
        };
        const response = await this.queryDatabase(databaseId, params);
        const {
          results: pages,
          next_cursor: nextCursor,
        } = response;

        for (const page of pages) {
          yield page;
        }

        cursor = nextCursor;

      } while (cursor);
    },
```

To reduce the number of API requests and optimize for [rate limiting](https://developers.notion.com/reference/request-limits#rate-limits), I think we can use the same approach for the `new-page` (and `new-database`) sources by replacing `last_edited_time` with `created_time`.

Reply:
Nice approach